### ISSUE TITLE
Exposing nested public symbols in Module

### ISSUE THREAD

**Issue by bkotsopoulossc** [emscripten-core/emscripten#15474]
Let's say I am using embind to define some bindings, and I have two different classes by the same name which I would like to expose:
```
namespace some_namespace1 { 
class Logger {
public:
  void log() {}
}; 
}
namespace some_namespace2 {
class Logger {
public:
  void log() {}
}; 
}

EMSCRIPTEN_BINDINGS(MyBindings) {
    em::class_<::some_namespace1::Logger>("Logger")
        .function("log", ::some_namespace1::Logger::log)
        ;
    em::class_<::some_namespace2::Logger>("Logger")
        .function("log", ::some_namespace2::Logger::log)
        ;
}
```

If we do this, we get a runtime error, `Cannot register public name 'Logger' twice`, from [here](https://github.com/emscripten-core/emscripten/blob/0c5046e64a80038bece746e3fde832d3ec1e9401/src/embind/embind.js#L1983), when registering the class, because the `.Logger` property is already defined on the Module, and it is not a function overload with a different number of arguments.

This is solvable by adding a prefix when binding each class, like `em::class_<::some_namespace1::Logger>("some_namespace1_Logger")`. 

But I wonder if we could instead support nesting exposed symbols into sub-objects at the Module level. So you could do something like:

```
    em::class_<::some_namespace1::Logger>("some_namespace1.Logger")
        .function("log", ::some_namespace1::Logger::log)
        ;
    em::class_<::some_namespace2::Logger>("some_namespace2.Logger")
        .function("log", ::some_namespace2::Logger::log)
        ;
```

We'd need to accept the dot character in the legal name check [here](https://github.com/emscripten-core/emscripten/blob/0c5046e64a80038bece746e3fde832d3ec1e9401/src/embind/embind.js#L2489).

We'd also need some logic [here](https://github.com/emscripten-core/emscripten/blob/0c5046e64a80038bece746e3fde832d3ec1e9401/src/embind/embind.js#L145-L148) to recursively apply this to the Module, by something like these [suggestions](https://stackoverflow.com/questions/20424226/easy-way-to-set-javascript-object-multilevel-property).

---

I wonder if this is something we'd want to support? Are there places where this feature would make things very difficult?

**Comment by sbc100** [comment]
Seems reasonable to me.  

But this also seems like a relatively niche case that as a known workaround which is to include some prefix into one or both of the conflicting symbols (i.e. `some_namespace1_Logger`).    Is specifying `"some_namespace1.Logger"` is somehow better then specifying `some_namespace1_Logger`?

**Comment by sbc100** [comment]
cc @RReverser

**Comment by bkotsopoulossc** [comment]
Yeah good question, like if there is such an easy workaround, why bother.

One nice part is it allows you to pass only part of Module around to functions, so it is clear what exposed symbols these functions use.

Let's say you have a couple classes in a couple namespaces, each with their own namespaces in c++, so your exposed symbols end up looking like this at the Module level:

```
Module.some_namespace1.Logger
Module.some_namespace1.Metrics
Module.some_namespace2.Logger
Module.some_namespace2.Database
...
```

Rather than always accessing Module from the global scope, you could pass it around, but pass a specific property (namespace in this case, but this property could be any group of types) that contains only certain types exposed as properties.

```
function initializeComponent1(namespace1Types) {
    namespace1Types.Logger.log(...)
    namespace1Types.Metrics.count(...)
}

initializeComponent1(Module.some_namespace1)
```

It also lends itself much nicer to writing typescript type definitions that match the structure of the exported symbols:

```
export interface Namespace1 {
  logger: Logger
  metrics: Metrics
}

export interface Module {
    namespace1: Namespace1
}

function initializeComponent1(namespace1Types: Namespace1) { ... }
```

And it potentially makes it simpler when there can be a 1:1 match with c++ type names, rather than adding long prefixes to match deeply nested namespaces.

**Comment by sbc100** [comment]
I think if you would like to implement it I sounds like reasonable feature and would likely be accepted. 

Still waiting to hear from @RReverser on this.

**Comment by RReverser** [comment]
To be honest, I don't have a strong opinion on this. I guess it doesn't harm if it doesn't result in too much extra complication to Embind, but at the same time I agree with earlier comments that this is a relatively niche case.

**Comment by bkotsopoulossc** [comment]
Thanks for the discussion.

What's the thoughts on having an API in embind for this. So something like
```
emscripten::class_<::some_namespace1::Logger>("Logger")
    .function("log", ::some_namespace1::Logger::log)
    ;
```

could be written out as
```
emscripten::class_<::some_namespace1::Logger>("Logger")
    .namespace("some_namespace1")
    .function("log", ::some_namespace1::Logger::log)
    ;
```

**Comment by finalpatch** [comment]
The registration happens in in the constructor of `class_` so the namespace name has to be passed to the constructor:

```
emscripten::class_<::some_namespace1::Logger>("some_namespace1.Logger")
    .function("log", ::some_namespace1::Logger::log)
    ;
```

The this requires changing a bunch of c++/js code (exposePublicSymbol, replacePublicSymobl, etc.) in embind to take into account that the name could be namespace qualified.

If we are okay with registering at module level, but then add an alias, we can have something like this:

```
emscripten::class_<::some_namespace1::Logger>("some_namespace1_Logger")
    .alias("some_namespace1.Logger")
    .function("log", ::some_namespace1::Logger::log)
    ;
```

Which should be fairly easy to add with minimal impact to existing code.
